### PR TITLE
Fix show results for multiple groups in /test/overview

### PR DESCRIPTION
**Pull request by ilausuch**
When the query has different groupid and every group
different builds the current operation only returns the jobs
for the last build of the first group.

This PR ensures to return all the jobs for all the last builds for
all the selected groups

https://progress.opensuse.org/issues/91650

### CODE REVIEW

**Comment by ilausuch** [comment]
Remains the test

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm [hunk 1]
```diff
@@ -382,11 +382,17 @@ sub _compose_job_overview_search_args {
...
     if (!$search_args{build}) {
-        # yield the latest build of the first group if (multiple) groups but not build specified
-        # note: the search arg 'groupid' is ignored by complex_query() because we later assign 'groupids'
-        $search_args{groupid} = $groups[0]->id if (@groups);
-
-        $search_args{build} = $schema->resultset('Jobs')->latest_build(%search_args);
+        if (@groups){
+            my @builds;
+            @builds = map {
+                $search_args{groupid} = $_->{_column_data}->{id};
```
Comment (perlpunk) [inline lib/OpenQA/WebAPI/Plugin/Helpers.pm:388]: Please don't use internal data of objects, especially not those which have an underscore in front :)

```suggestion
                $search_args{groupid} = $_->id;
```

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm [hunk 2]
```diff
@@ -382,11 +382,17 @@ sub _compose_job_overview_search_args {
...
+                $schema->resultset('Jobs')->latest_build(%search_args);
+            } @groups;
+            $search_args{build} = \@builds;
```
Comment (perlpunk) [inline lib/OpenQA/WebAPI/Plugin/Helpers.pm:391]: Some of the builds (or even all) gathered in the previous `map` might be undef.
```suggestion
            $search_args{build} = [grep { defined $_ } @builds];
```

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 1 above)
Comment (ilausuch) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm:388]: I think I tested this before and didn't work. I'll check it again

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 2 above)
Comment (perlpunk) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm:391]: In my test I got `[ undef, ':TW.8318' ]` in `@builds`.

Yes, conditionally returning `()` would also work.

e.g.
```perl
$schema->resultset('Jobs')->latest_build(%search_args) || ();
```

One could even save another statement:
```perl
my @builds = map {
    $schema->resultset('Jobs')->latest_build(%search_args, groupid => $_->id) || ();
} @groups;
```

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 1 above)
Comment (ilausuch) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm:388]: I am not sure, but this doesn't work for me. the test runs but not in a browser

**Comment by ilausuch** [comment]
An other interesting thing is that in the test now I have the same build duplicated 0091,0091 so I should introduce a "distinct" of build and not include duplicates

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 1 above)
Comment (perlpunk) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm:388]: What do you mean with "it doesn't work"?
It works for me, I get the expected result.
Do you get an error message?
Did you trace the SQL statements with DBIC_TRACE?

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 2 above)
Comment (perlpunk) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm]: To avoid duplicated build names (as you mentioned below) I suggest the following:
```perl
            my %builds;
            for my $group (@groups) {
                my $build = $schema->resultset('Jobs')->latest_build(%search_args, groupid => $group->id) or next;
                $builds{ $build }++;
            }
            $search_args{build} = [sort keys %builds];
```

File: lib/OpenQA/WebAPI/Controller/Test.pm [hunk 3]
```diff
@@ -695,6 +694,8 @@ sub overview {
         }
     }
 
+
+
```
Comment (okurz) [inline lib/OpenQA/WebAPI/Controller/Test.pm:698]: unrelated whitespace changes, please revert

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 2 above)
Comment (ilausuch) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm]: done

File: lib/OpenQA/WebAPI/Plugin/Helpers.pm (hunk 1 above)
Comment (ilausuch) [reply lib/OpenQA/WebAPI/Plugin/Helpers.pm]: Solved with your next suggestion

File: lib/OpenQA/WebAPI/Controller/Test.pm (hunk 3 above)
Comment (ilausuch) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: done

File: t/10-tests_overview.t [hunk 4]
```diff
@@ -233,8 +233,8 @@ $t->get_ok('/tests/overview?distri=opensuse&version=13.1&groupid=1001&groupid=10
 $summary = get_summary;
 like(
     $summary,
-    qr/Summary of opensuse, opensuse test/i,
-    'multiple groups with no build specified yield latest build of first group'
+    qr/Summary of opensuse, opensuse test build 0091 /i,
```
Comment (ilausuch) [inline t/10-tests_overview.t:236]: This extra space is to ensure we don't have more than one build in the string

File: t/10-tests_overview.t [hunk 5]
```diff
@@ -233,8 +233,8 @@ $t->get_ok('/tests/overview?distri=opensuse&version=13.1&groupid=1001&groupid=10
...
+    'multiple groups with no build specified yield latest build of every group'
```
Comment (perlpunk) [inline t/10-tests_overview.t:237]: It would probably make sense to add a test with groups that result in two different build names. Right now the result is like before, so the test doesn't prove that the change works.

Comment (ilausuch) [reply t/10-tests_overview.t:237]: Done

File: t/10-tests_overview.t [hunk 6]
```diff
@@ -229,14 +229,23 @@ $t->element_exists('#res_DVD_i586_kde',                           'job from grou
...
+# $t->get_ok('/tests/overview?distri=opensuse&version=13.1&groupid=1001&groupid=1002')->status_is(200);
+# $summary = get_summary;
+# like(
+#     $summary,
+#     qr/Summary of opensuse, opensuse test build 0091 /i,
+#     'multiple groups with no build specified yield latest build of every group'
+# );
+# like($summary, qr/Passed: 2 Failed: 0 Scheduled: 1 Running: 2 None: 1/i);
+
 $t->get_ok('/tests/overview?distri=opensuse&version=13.1&groupid=1001&groupid=1002')->status_is(200);
```
Comment (okurz) [inline t/10-tests_overview.t:241]: Instead of adding the job to the lengthy fixture list breaking many other test modules better just create a single job before you do the "get", e.g.

```perl
$t->app->schema->resultset('Jobs')->create({
        id => 99964,
        BUILD => '0092',
        group_id => 1002,
        TEST => 'kde'
    });
```

**Review by okurz — CHANGES_REQUESTED** [review]

File: t/10-tests_overview.t (hunk 4 above)
Comment (kalikiana) [reply t/10-tests_overview.t:236]: Why not use `$` there?

File: lib/OpenQA/WebAPI/Controller/Test.pm [hunk 7]
```diff
@@ -673,15 +673,14 @@ sub overview {
...
-        build   => $search_args->{build},
+        build   => ref $search_args->{build} eq 'ARRAY' ? join(',', @{$search_args->{build}}) : $search_args->{build},
         version => $search_args->{version},
         distri  => $search_args->{distri},
         groups  => $groups,
         until   => $until,
     );
     my @latest_jobs = $self->schema->resultset('Jobs')->complex_query(%$search_args)->latest_jobs($until);
     ($stash{archs}, $stash{results}, $stash{aggregated}) = $self->prepare_job_results(\@latest_jobs);
-
```
Comment (okurz) [inline lib/OpenQA/WebAPI/Controller/Test.pm:684]: please remove the unrelated whitespace change

File: t/10-tests_overview.t (hunk 4 above)
Comment (okurz) [reply t/10-tests_overview.t:236]: Because then the regex shouldn't match

**Review by okurz — CHANGES_REQUESTED** [review]
Please squash the commits. The test should be part of the same commit as the implementation

File: t/10-tests_overview.t [hunk 8]
```diff
@@ -233,11 +233,40 @@ $t->get_ok('/tests/overview?distri=opensuse&version=13.1&groupid=1001&groupid=10
...
+        DISTRI   => 'opensuse',
+        VERSION  => '13.1'
+    });
+
+$t->get_ok('/tests/overview?distri=opensuse&version=13.1&groupid=1001&groupid=1003')->status_is(200);
+$summary = get_summary;
+like(
+    $summary,
+    qr/Summary of opensuse, opensuse test 2 build 0091,0092 /i,
+    'multiple groups with no build specified yield latest build of every group'
```
Comment (perlpunk) [inline t/10-tests_overview.t:263]: The test label is now the same as the previous one.
Labels should be different for each test, so that it is easy to find out which test failed.

File: t/10-tests_overview.t (hunk 4 above)
Comment (kalikiana) [reply t/10-tests_overview.t:236]:
```suggestion
    qr/Summary of opensuse, opensuse test build 0091[^,]/i,
```

Comment (okurz) [reply t/10-tests_overview.t:236]: yes, that works. I have created https://github.com/os-autoinst/openQA/pull/3973 incorporating the latest review suggestions from here so please approve there as long as ilausuch did not update their PR

File: t/10-tests_overview.t (hunk 8 above)
Comment (okurz) [reply t/10-tests_overview.t:263]: incorporated into https://github.com/os-autoinst/openQA/pull/3973

File: lib/OpenQA/WebAPI/Controller/Test.pm (hunk 7 above)
Comment (okurz) [reply lib/OpenQA/WebAPI/Controller/Test.pm:684]: included in https://github.com/os-autoinst/openQA/pull/3973

**Comment by ilausuch** [comment]
incorporated into #3973
